### PR TITLE
fix bug in key condition

### PR DESCRIPTION
**Pull request by hanfei1991**
It's not right to cache result column in `apply function` only by function name, because the params may be different.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
fix bug in key condition

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit c2f6751dafe3ecd0841f85799c0b15ae91e7cef9 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68354/c2f6751dafe3ecd0841f85799c0b15ae91e7cef9/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by jsc0218 — APPROVED** [review]

**Comment by hanfei1991** [comment]
It seems `applyFunction` will calculate and cache a column, rather than a field .... then it will be slow when there are many marks ... 

Can we use `function ptr to string` to replace `function name` for cache?

File: src/Storages/MergeTree/KeyCondition.cpp [hunk 1]
```diff
@@ -894,7 +894,11 @@ static FieldRef applyFunction(const FunctionBasePtr & func, const DataTypePtr &
     if (field.isExplicit())
         return applyFunctionForField(func, current_type, field);
 
-    String result_name = "_" + func->getName() + "_" + toString(field.column_idx);
+    WriteBufferFromOwnString buf;
+    writeText("_", buf);
+    writePointerHex(func.get(), buf);
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/KeyCondition.cpp:899]: Sounds insecure.

Comment (alexey-milovidov) [reply src/Storages/MergeTree/KeyCondition.cpp:899]: Missing comments.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Our style guide mandates writing comments in the code.
This cannot be merged without comments.

**Comment by jsc0218** [comment]
> It seems `applyFunction` will calculate and cache a column, rather than a field .... then it will be slow when there are many marks ...
> 
> Can we use `function ptr to string` to replace `function name` for cache?
> 
> Edit: it looks like workable! @jsc0218 PTAL

I am not sure as I thought the same with you at the first glance, probably others can take a look for this point.

File: src/Storages/MergeTree/KeyCondition.cpp (hunk 1 above)
Comment (CurtizJ) [reply src/Storages/MergeTree/KeyCondition.cpp:904]: This function is only used in `applyMonotonicFunctionsChainToRange`. So you can just pass the index of function from [array](https://github.com/ClickHouse/ClickHouse/blob/008d02880bdf99fa659d34790d5ed1e9b884b6d8/src/Storages/MergeTree/KeyCondition.cpp#L2543) as a column name.

Comment (hanfei1991) [reply src/Storages/MergeTree/KeyCondition.cpp:904]: but I think the lenth of the functions should be 1.
elements `intDiv(a, 20)` and `intDiv(a, 50000)` have there own `monotonic functions chain` which length is 1.

Comment (hanfei1991) [reply src/Storages/MergeTree/KeyCondition.cpp:904]: @alexey-milovidov @CurtizJ PTAL, we need this fix

Comment (CurtizJ) [reply src/Storages/MergeTree/KeyCondition.cpp:904]: Well, you are right. Maybe using a pointer as a name is not that bad.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**